test-data/unit/check-functions.test

```diff
@@ -734,6 +734,188 @@ a = None # type: A
+    def f(self): pass
+
+foo = Foo()
+
+foo.f("a")  # E: Argument 1 to "f" of "Foo" has incompatible type "str"; expected "int"
+x: str = foo.f(1)
+y: int = foo.f(1)  # E: Incompatible types in assignment (expression has type "str", variable has type "int")
+
+reveal_type(foo.f)  # E: Revealed type is 'def (builtins.int) -> builtins.str'
+
+[builtins fixtures/dict.pyi]
+
+[case testClassMethodWithDeclaredDecoratedType]
+
+from typing import Callable, Any
+from mypy_extensions import declared_type
+
+def dec(f): pass
+
+# Note that the decorated type must account for the `cls` argument  -- It's applied pre-binding
+class Foo:
+    @declared_type(Callable[[Any, int], str])
+    @classmethod
+    @dec
+    def f(cls): pass
+
+
+Foo.f("a")  # E: Argument 1 to "f" of "Foo" has incompatible type "str"; expected "int"
+x: str = Foo.f(1)
+y: int = Foo.f(1)  # E: Incompatible types in assignment (expression has type "str", variable has type "int")
+
+reveal_type(Foo.f)  # E: Revealed type is 'def (builtins.int) -> builtins.str'
+
+[builtins fixtures/dict.pyi]
+
+[case testStaticMethodWithDeclaredDecoratedType]
+
+from typing import Callable
+from mypy_extensions import declared_type
+
+def dec(f): pass
+
+class Foo:
+    @declared_type(Callable[[int], str])
+    @staticmethod
+    @dec
+    def f(): pass
+
+
+Foo.f("a")  # E: Argument 1 to "f" of "Foo" has incompatible type "str"; expected "int"
+x: str = Foo.f(1)
+y: int = Foo.f(1)  # E: Incompatible types in assignment (expression has type "str", variable has type "int")
+
+reveal_type(Foo.f)  # E: Revealed type is 'def (builtins.int) -> builtins.str'
+
+[builtins fixtures/dict.pyi]
+
+[case testUntypedDecoratorWithDeclaredType]
+
+from typing import Callable
+from mypy_extensions import declared_type
+
+def dec(f): pass
+
+@declared_type(Callable[[int], str])
+@dec
+def f(): pass
+
+f("a")  # E: Argument 1 to "f" has incompatible type "str"; expected "int"
+x: str = f(1)
+y: int = f(1)  # E: Incompatible types in assignment (expression has type "str", variable has type "int")
+
+reveal_type(f)  # E: Revealed type is 'def (builtins.int) -> builtins.str'
+
+[builtins fixtures/dict.pyi]
+
+[case testDecoratorWithDeclaredTypeBare]
+
+from typing import Callable
+from mypy_extensions import declared_type
+
+@declared_type(Callable[[int], str])
+def f(x): pass
+
+f("a")  # E: Argument 1 to "f" has incompatible type "str"; expected "int"
+x: str = f(1)
+y: int = f(1)  # E: Incompatible types in assignment (expression has type "str", variable has type "int")
+reveal_type(f)  # E: Revealed type is 'def (builtins.int) -> builtins.str'
+
+[builtins fixtures/dict.pyi]
+
+[case testDecoratorWithDeclaredTypeIncompatible]
+
+from typing import Callable, TypeVar
+from mypy_extensions import declared_type
+
+T = TypeVar('T')
+
+def dec(f: T) -> T: pass
+
+@declared_type(Callable[[int], str])  # E: Incompatible types (inferred decorated type Callable[[str], str], declared decorated type Callable[[int], str])
+@dec
```

Comment:
Also test `declared_type` together with another decorator that modifies the signature of the function. Test both a good usage and an invalid usage (where the declared type is incompatible).